packages/core/src/application.ts

```diff
@@ -3,555 +3,54 @@
-import {getControllerSpec} from './router/metadata';
-import {HttpHandler} from './http-handler';
-import {writeResultToResponse} from './writer';
-import {DefaultSequence, SequenceHandler, SequenceFunction} from './sequence';
-import {RejectProvider} from './router/providers/reject';
-import {
-  FindRoute,
-  InvokeMethod,
-  Send,
-  Reject,
-  ParseParams,
-} from './internal-types';
-import {ControllerClass} from './router/routing-table';
-import {GetFromContextProvider} from './router/providers/get-from-context';
-import {BindElementProvider} from './router/providers/bind-element';
-import {InvokeMethodProvider} from './router/providers/invoke-method';
-import {FindRouteProvider} from './router/providers/find-route';
-import {CoreBindings} from './keys';
-
-const SequenceActions = CoreBindings.SequenceActions;
-
-// NOTE(bajtos) we cannot use `import * as cloneDeep from 'lodash/cloneDeep'
-// because it produces the following TypeScript error:
-//  Module '"(...)/node_modules/@types/lodash/cloneDeep/index"' resolves to
-//  a non-module entity and cannot be imported using this construct.
-const cloneDeep: <T>(value: T) => T = require('lodash/cloneDeep');
-
-const debug = require('debug')('loopback:core:application');
-
-interface OpenApiSpecOptions {
-  version?: string;
-  format?: string;
-}
-
-const OPENAPI_SPEC_MAPPING: { [key: string] : OpenApiSpecOptions; } = {
-  '/openapi.json': {version: '3.0.0', format: 'json'},
-  '/openapi.yaml': {version: '3.0.0', format: 'yaml'},
-  '/swagger.json': {version: '2.0', format: 'json'},
-  '/swagger.yaml': {version: '2.0', format: 'yaml'},
-};
+import { Context, Binding, BindingScope } from '@loopback/context';
+import { Server } from './server';
 
 export class Application extends Context {
-  /**
-   * Handle incoming HTTP(S) request by invoking the corresponding
-   * Controller method via the configured Sequence.
-   *
-   * @example
-   *
-   * ```ts
-   * const app = new Application();
-   * // setup controllers, etc.
-   *
-   * const server = http.createServer(app.handleHttp);
-   * server.listen(3000);
-   * ```
-   *
-   * @param req The request.
-   * @param res The response.
-   */
-  public handleHttp: (req: ServerRequest, res: ServerResponse) => void;
-
-  protected _httpHandler: HttpHandler;
-  protected get httpHandler(): HttpHandler {
-    this._setupHandlerIfNeeded();
-    return this._httpHandler;
-  }
-
-  constructor(public options?: ApplicationOptions) {
+  // Keep a local reference here to simplify iteration over registered
+  // servers at app start.
+  public servers: Map<string, Server>;
+  constructor() {
     super();
-
-    if (!options) options = {};
-
-    this.bind(CoreBindings.HTTP_PORT).to(
-      options.http ? options.http.port : 3000,
-    );
-    this.api(createEmptyApiSpec());
-
-    if (options.components) {
-      for (const component of options.components) {
-        this.component(component);
-      }
-    }
```

Comment:
Components is a concept that's an integral part of `core`, we need it to register extensions. For example, in an application wishing to use legacy juggler bridge for data access, it should be enough to register `RepositoryComponent` once in the app, users should not need to register the component for each server/transport they are using.

Reply:
Do we want components in both `Application` and `Server`?

Reply:
That's a good question, let's discuss it in https://github.com/strongloop/loopback-next/pull/576#discussion_r139726378.